### DISCUSSION TITLE
Can't use "<" or ">" on the command line

### DISCUSSION THREAD

**Discussion by Fernando-Cabral** [xonsh/xonsh#5117] [General]
Question: how can I pass "<X" as a command line option to a program without seeing xonsh intepreting it as a redirect operattor?

I have a command line parameter like this:
```xsh
showcmd -d "<20"
showcmd -d ">20"
showcmd -d "=20"

showcmd "<"
showcmd ">"

```
These < and > are not supposed to be redirect, they are just command line parameter. Nevertheless, xonsh insists in taking them as a file to read from as in <20 or a file to write to, as in >20.
I tried to protect the symbols < and > from being interpreted by xonsh to no avail. I tried "<", '<', `<` and even \< (I know this last one shouln't work, but I tried it anyway. )
So, I can not run this particular command because I have found no way to pass on the parameter it expects.

Is there a way to go around this?

**Comment by anki-code** [comment]
This looks like issue.
```xsh
showcmd "<20"
# ['<20']  

# works as expected, but...

showcmd "<"
# xonsh: None: unable to open file

showcmd @("<")
# xonsh: None: unable to open file

showcmd $(echo -n '<')
# xonsh: None: unable to open file

showcmd! <
# xonsh: None: unable to open file

@(['showcmd', '<'])
# xonsh: None: unable to open file

# Expected:
# ['<']
```
May be the related PR was - https://github.com/xonsh/xonsh/pull/4826

**Comment by Fernando-Cabral** [comment]
Yes, it pretty much looks like this issue is related to the [#4826](https://github.com/xonsh/xonsh/pull/4826) issue you referred to. Anyway, as I see it, no expansion should be attempted when symbols or characters are quoted either with a single or double quotes.

```
cmd "<"
cmd '>'
cmd ">"
cmd '<'
```
Both should work the same.
In this particular case it should work even if using back ticks, provided no file satisfies the expression, as in:
```
cmd `>`
cmd `<`
```
But alas, presently, none of them works as expected.
Work-around for the time being: I put a .strip() before using the parameter and instructed the user to make sure he puts an space before the symbol, like so:
```
cmd " >" " <"
```
It works, which means xonsh does not try to redirect when "<" and ">" are not the first character in the quoted string.

**Reply by jbw3** [reply]
I'm not sure this is related to my PR #4826. I tested Xonsh 0.13.0 (before my PR got merged in), and I am still seeing this issue.

**Comment by bestlem** [comment]
Doesn't ~ need to be expanded if inside quotes? If not thet gets to be a real pain.

Thus some expansion/parsing has to be done.

**Reply by Fernando-Cabral** [reply]
I agree that could be an issue.
But then we could construe a convention that what is inside double quotes would be expanded but what is inside single quotes wouldn't. (I think this is what every regular shell does).
Or perhaps the least problematic solution would be to make \` (back tick) to work as expected. I don't see any reason to have \`<\` intepreted as a stdin redirection. So, in this case, something like:
```
cmd `<5`
cmd `>1`
```
would work except if there are files named "<" or "<", which would never a good idea anyway. (I can't imagine any sane person naming a file ">5"...)

**Reply by anki-code** [reply]
The back quotes in xonsh is for regex pattern - https://xon.sh/tutorial.html#regular-expression-globbing

**Reply by Fernando-Cabral** [reply]
I understand that in xonsh back quotes is for regex. The problem is that it is not working as such. If it were, the command
```
cmd `<`
```
would just be translated into `cmd <`. That is, xonsh would "expand" the regular expression \`<\` to its literal value, <. In this case `cmd` would never know there was a back quote in the command line. The same would happen if the parameter was a little bit different, as is in the real case:
```
cmd `<2`
cmd `>2`
cmd `=2`
```
Unfortunately xonsh does not understand that "<2", '<2' or \`<2\` should not be intepreted as `<2`. It takes `2` as a file name and \`<\` a redirection where a file name is missing. See the error messages:
```
cmd `<`
cmd `<a`

Unrecognized redirection command: <a
```
So, any way you look at it, xonsh is wrong.
What I said was quite simple: if xonsh sees \`<\`  as  regular expression the command above would work EVEN THOU back quotes are used for a non-cannonical purpose.

**Reply by anki-code** [reply]
Could you please provide real life example of using single angle brackets as argument and create the issue?

**Reply by Fernando-Cabral** [reply]
Sure. Every example I gave above is real. The program I am running takes several arguments that can start with one of the following: `>, <, =, +, -`.  It is a utility that has some similarities with the classic `find`. Those prefixes are usually applied to numbers. For instance, when searching for files with certain attributes:
```
find --links  >10    # Files with more than 10 links
find --links -10      # Except files with 10 links
find --links <10     # Files with less than 10 links 
find --size <100k  # Files with less than 100k
find --date >2022-12-01  # Files created after 2022-12-01
find --date <2022-12-01  # Files created before 2022-12-01
find --date -2022-12-01  # Files except those created in 2022-12-01
``` 
With xonsh - but not with any other classic shells - it has been close o impossible to use the prefixes "<" and ">".

(As I said, the trick I have been using is to add an white space as in `" <10"`. To do this, I had to force the user to prepend a white space and make sure the program trims it off before processing the arguments.)

**Reply by bestlem** [reply]
I thought I had posted this before but Github seems to have swallowed it.

    echo ">"

results in 

    xonsh: None: unable to open file

**Reply by Fernando-Cabral** [reply]
It was there since yesterday, but now I only see this new post.
Weird.
